fem/hyperbolic_conservation_laws.cpp

```diff
@@ -0,0 +1,493 @@
+      elfun1_mat.MultTranspose(shape1, state1);
+      elfun2_mat.MultTranspose(shape2, state2);
+
+      // Get the normal vector and the flux on the face
+      if (nor.Size() == 1)  // if 1D, use 1 or -1.
+      {
+         // This assume the 1D integration point is in (0,1). This may not work if
+         // this chages.
+         nor(0) = (Tr.GetElement1IntPoint().x - 0.5) * 2.0;
+      }
+      else
+      {
+         CalcOrtho(Tr.Jacobian(), nor);
+      }
+      // Compute F(u+, x) and F(u-, x) with maximum characteristic speed
+      // Compute hat(F) using evaluated quantities
+      const double speed = rsolver.Eval(state1, state2, nor, Tr, fluxN);
+
+      // Update the global max char speed
+      max_char_speed = std::max(speed, max_char_speed);
+
+      // pre-multiply integration weight to flux
+      fluxN *= ip.weight;
+      for (int k = 0; k < num_equations; k++)
+      {
+         // this loop structure can increase cache hit because
+         for (int s = 0; s < dof1; s++)
+         {
+            elvect1_mat(s, k) -= fluxN(k) * shape1(s);
+         }
+         for (int s = 0; s < dof2; s++)
+         {
+            elvect2_mat(s, k) += fluxN(k) * shape2(s);
+         }
+      }
+   }
+}
+
+HyperbolicFormIntegrator::HyperbolicFormIntegrator(
+   const FluxFunction &fluxFunction, const RiemannSolver &rsolver,
+   const int IntOrderOffset)
+   : NonlinearFormIntegrator(),
+     fluxFunction(fluxFunction),
+     IntOrderOffset(IntOrderOffset),
+     rsolver(rsolver),
+     num_equations(fluxFunction.num_equations)
+{
+#ifndef MFEM_THREAD_SAFE
+   state.SetSize(num_equations);
+   flux.SetSize(num_equations, fluxFunction.dim);
+   state1.SetSize(num_equations);
+   state2.SetSize(num_equations);
+   fluxN1.SetSize(num_equations);
+   fluxN2.SetSize(num_equations);
+   fluxN.SetSize(num_equations);
+   nor.SetSize(fluxFunction.dim);
+#endif
+}
+
+double FluxFunction::ComputeFluxDotN(const Vector &U,
+                                     const Vector &normal,
+                                     FaceElementTransformations &Tr, Vector &FUdotN) const
+{
+   double val = ComputeFlux(U, Tr, flux);
+   flux.Mult(normal, FUdotN);
+   return val;
+}
+
+
+
+void DGHyperbolicConservationLaws::Update()
+{
+   nonlinearForm->Update();
+   height = nonlinearForm->Height();
+   width = height;
+   z.SetSize(height);
+
+   ComputeInvMass();
+}
+double RusanovFlux::Eval(const Vector &state1, const Vector &state2,
+                         const Vector &nor, FaceElementTransformations &Tr,
+                         Vector &flux) const
+{
+#ifdef MFEM_THREAD_SAFE
+   Vector fluxN1(fluxFunction.num_equations), fluxN2(fluxFunction.num_equations);
+#endif
+   const double speed1 = fluxFunction.ComputeFluxDotN(state1, nor, Tr, fluxN1);
+   const double speed2 = fluxFunction.ComputeFluxDotN(state2, nor, Tr, fluxN2);
+   // NOTE: nor in general is not a unit normal
+   const double maxE = std::max(speed1, speed2);
+   flux = state1;
```

Comment:
Purely a style question, but any reason to split up the flux calculation across multiple lines instead of having a one liner?